Allow mounting at non-root path, e.x. foo.com/metabase/, v2

tlrobinson opened the pull request:
Resolves #2090 

TODO:

* [x] react-router
* [x] API calls
* [x] CSS `url()`
* [x] `<img>`s
* [x] inject `<base>` tag for assets
* [x] set cookie path
* [x] ensure none of the variables injected in the HTML are vulnerable to XSS
* ~docs~

Built jar (as of f1e317d): https://www.dropbox.com/s/bpnbse7ndw655zv/metabase-pr-4740.jar?dl=1

Example Caddyfile (name it `Caddyfile` then run `caddy`):

```
http://localhost:3100

proxy /metabase localhost:3000 {
  without /metabase
}
```

Example Nginx config:

```
server {
   listen       3100;
   server_name  localhost;

   location /metabase/ {
       proxy_pass http://localhost:3000/;
   }
}
```

tlrobinson committed:
Allow running at non-root path, version 2. Resolves #2090

tlrobinson committed:
Fix image/geojson assets, and consolidate them

tlrobinson committed:
Improved root path handling. Uses path in site-url, but checks to make sure it's correct.

luc-poynt commented:
On it : brand new installation and im stuck in setup, it keep send me back to setup,
created my use account, will add data later.. and back to the setup loop....

tlrobinson commented:
@luc-poynt Cool!

I'm not exactly sure about your remark. When you run Metabase it should use `metabase.db.mv.db`  in the current directory (or whatever you set `MB_DB_FILE` to) as the application database and you shouldn't need to set up the instance again. Maybe you ran it from a different directory and it created a new database?

Also, FYI we recommend using Postgres or MySQL for the application database in production.

luc-poynt commented:
@tlrobinson 

The init script is set to use mysql, database created, user can access the db no issue.

So the very first time Metabase runs is goes to couple step (setup), one of them is setup 'data'
and choose I'll do this later.. so then Im getting setup loop :) 
Once I choose to set it and  use Mysql all works.

make sense ?

tlrobinson committed:
Merge branch 'master' of github.com:metabase/metabase into non-root-path-v2

tlrobinson commented:
@luc-poynt Yeah, I see what you mean. Looking into it.

tlrobinson committed:
Migrate a couple fetch calls to services, fix some <a href="#"> links.

tlrobinson commented:
@luc-poynt Fixed that issue, updated the jar linked above.

tlrobinson committed:
Set cookie path to MetabaseRoot

wvengen commented:
Thanks! The assets are fetched from the non-root path, that looks good. But Metabase doesn't work:
```
Refused to execute script from '...' because its MIME type ('text/html') is not executable, and strict MIME type checking is enabled.
```

Happens both on Chromium and Firefox.

dncpax commented:
@tlrobinson Hi there. Great news!
IT seems sharing is broken because it generates links that do not respect the root:
http://server/public/question/UUID
It should use the proxyed location:
http://server/metabase/public/question/UUID

Embeding uses the same url.

Have not tested sharing in application code. Same thing can be happening there too.

luc-poynt commented:
For reference, here is my nginx config
upstream metabase_server {
    server 127.0.0.1:3000 fail_timeout=0;
}

server {
    listen 80;
    server_name metabase.momo.net;
    return 301 https:://metabase2.momo.net:$request_uri;
}

server {
    listen 443;
    server_name metabase2.momo.net;

    include /etc/nginx/modules.d/ssl.conf;

    ignore_invalid_headers off; #pass through headers from metabase which are considered invalid by Nginx server.
    location /metabase/ {
        sendfile off;
        proxy_pass      http://metabase_server/;
        proxy_redirect  default;

        proxy_set_header   Host             $host;
        proxy_set_header   X-Real-IP        $remote_addr;
        proxy_set_header   X-Forwarded-For  $proxy_add_x_forwarded_for;
        proxy_max_temp_file_size 0;

        # this is the maximum upload size
        client_max_body_size    10m;
        client_body_buffer_size 128k;

        proxy_connect_timeout   90;
        proxy_send_timeout      90;
        proxy_read_timeout      90;

        proxy_buffer_size           4k;
        proxy_buffers               4 32k;
        proxy_busy_buffers_size     64k;
        proxy_temp_file_write_size  64k;
    }
}

tlrobinson commented:
@dncpax Have you tried updating your "site url" in the admin settings?

If you haven't, first can you tell me if you're seeing this warning https://github.com/metabase/metabase/pull/4740/files#diff-a2e456f81e57ed0d878a736fbc3c05f2R36 in the JavaScript console?

Then try updating the site url.

We detect the correct "base" name to make sure the app loads, even with a misconfigured "site url", but we don't use that detected base name for generating user-facing URLs. I *think* we should use the actual "site url" setting for that, since it's what the backend uses for pulses, etc.

dncpax commented:
@tlrobinson No, I hadn't updated site url.
I see the warning in the console:
Site URL root "/" does not match actual root "/mb2/"

Also, I see a few 404 errors fetching 2 css and 2 js:
general:43 GET http://192.168.2.25/app/dist/styles.bundle.css?028da4709a7476696aa98c22a40f637e 
general:43 GET http://192.168.2.25/app/dist/app-main.bundle.css?8143040f9501d55903ddc86b3ca4142d 
general:43 GET http://192.168.2.25/app/dist/styles.bundle.js?61cb0217b8828bb0495a 
general:43 GET http://192.168.2.25/app/dist/app-main.bundle.js?61cb0217b8828bb0495a 

After updating site url... no more errors or warnings in the console. But...

The share links are still incorrect.

maesenka commented:
I've also been testing this, and it works great. No problems so far.

We use it behing an NGINX proxy with the same config as in the example given by tlrobinson above.

dncpax commented:
@maesenka Did you check your sharing url's?

maesenka commented:
@dncpax You're right. The generated links don't work. The path element is missing.

dncpax commented:
Almost there definitely. The simple configuration is really nice. I can live with having to correct the shareable links. Just hope this makes it into v0.24!

tlrobinson committed:
Merge branch 'master' of github.com:metabase/metabase into non-root-path-v2

tlrobinson committed:
Merge branch 'master' of github.com:metabase/metabase into non-root-path-v2

tlrobinson committed:
Fix public link URLs

tlrobinson committed:
More helpful warning when site url isn't correct

dncpax commented:
open source rocks!

tlrobinson committed:
Escape strings we're injecting in the index.html script to prevent XSS.

tlrobinson commented:
Alright I fixed the public links issue, made the warning console.log a little clearer, and fixed a potential XSS vulnerability.

I think this is ready for review, and documentation, if desired.

As pointed out by @thearthur, we should document that if you choose to "mount" Metabase on a subdirectory of a domain shared with another site, Metabase will have complete access to that site's cookies and content, and vice versa. This inherent to the way browsers work.

tlrobinson committed:
Fix str name conflict

tlrobinson committed:
Fix docstring

tlrobinson committed:
Fix a few more absolute paths

attekei commented on frontend/src/metabase/home/components/Activity.jsx:
```diff
@@ -1,5 +1,4 @@
-import React, { Component } from 'react';
-import PropTypes from "prop-types";
+import React, { Component, PropTypes } from 'react';
```
This change is probably accidental as we are now using `prop-types`?

attekei commented on frontend/src/metabase/home/components/NewUserOnboardingModal.jsx:
```diff
@@ -64,11 +64,11 @@ export default class NewUserOnboardingModal extends Component {
                             <h2>Just 3 things worth knowing</h2>
 
-                            <p className="clearfix pt1"><img className="float-left mr2" width="40" height="40" src="/app/home/partials/onboarding_illustration_tables.png" />All of your data is organized in Tables. Think of them in terms of Excel spreadsheets with columns and rows.</p>
+                            <p className="clearfix pt1"><img className="float-left mr2" width="40" height="40" src="app/assets/img/onboarding_illustration_tables.png" />All of your data is organized in Tables. Think of them in terms of Excel spreadsheets with columns and rows.</p>
```
👍  good that you got rid of that strange `partials` resource directory

attekei reviewed (changes requested):
Just that tiny `prop-types` thing, other than that looks good. I also tested with caddy that the app and embeds seemed to work normally.

Would writing a tiny E2E test for this be a good idea? It would change the root url, navigate to the new root url, see that the result isn't 404, and then change the root url back to normal so that other tests won't be  affected.

attekei commented:
The site url setting text `The base URL of this Metabase instance, e.g. "http://metabase.my-company.com".` could be updated to mention the support for non-root paths, probably just by giving an example would be enough

tlrobinson committed:
Merge branch 'master' of github.com:metabase/metabase into non-root-path-v2

tlrobinson commented:
@metabase/core-developers Do we even need to document this? Honestly I don't think we want to encourage this due to the potential security implications mentioned above. It seems like plenty of people have tried it without documentation, so I'm inclined to leave it undocumented.

We could show a warning in the settings when we detect your site-url doesn't match the actual URL, since otherwise various links will break.

tlrobinson committed:
stray proptypes

salsakran commented:
Honestly, given that it's generally not a good idea to do this, I don't think we should point people to it.

dncpax commented:
@tlrobinson @salsakran I really can't see why this is a bad practice? It's like any other site or am I missing something? Does using a web server with several apps make it specially insecure? I can see why it is preferable to isolate each app but that is not a feasible scenario in many if not most of the times.

tlrobinson commented:
It just means if one application on the domain is somehow compromised (via XSS or whatever) then all applications on the domain are. Maybe that's not a big deal in your situation.

People who want this feature will be glad it's there, but I don't think we need to encourage it.

tlrobinson committed:
Fix geojson test

tlrobinson committed:
Fix flow errors [ci all]

dncpax commented:
@tlrobinson not to take anymore of your time with this, but ins't there a limit of subdomains or server names you can manage? There must be a point where you prefer to have to have some apps that reside on a server under subfolders. I'm just making a case that this is not a fringe practice. Anyway you choose to document this, it's perfectly fine with me. Really glad you took this on.

tlrobinson committed:
Merge branch 'master' of github.com:metabase/metabase into non-root-path-v2 [ci e2e]

tlrobinson committed:
Merge branch 'master' of github.com:metabase/metabase into non-root-path-v2

tlrobinson commented:
@salsakran do we want to merge this for 0.24 or wait to 0.25?

salsakran commented on frontend/src/metabase/home/components/NextStep.jsx:
```diff
@@ -1,6 +1,6 @@
 import React, { Component } from "react";
 import { Link } from "react-router";
-import fetch from 'isomorphic-fetch';
+import { SetupApi } from "metabase/services";
```
👍 to not having a bunch of urls sprinkled through the codebase!

salsakran reviewed (approved):
LGTM.

We shouldn't really call attention to this, but good to remove the constraint of needing to be at the root.

camsaul commented on src/metabase/routes.clj:
```diff
@@ -13,6 +14,14 @@
             [stencil.core :as stencil]))
 
+(defn- base-href []
+  (str (.getPath (clojure.java.io/as-url (public-settings/site-url))) "/"))
```
SInce `clojure.java.io` is already required as `io` just use the alias, `(io/as-url ...)`

camsaul commented on src/metabase/routes.clj:
```diff
@@ -13,6 +14,14 @@
+(defn- base-href []
+  (str (.getPath (clojure.java.io/as-url (public-settings/site-url))) "/"))
+
+(defn- escape-script [text]
```
Probably makes sense to clarify this a bit and call it `escape-script-tags` or at the very least `escape-scripts`

Also in this case it's idiomatic Clojure to use `s` if you have a generic string parameter, so I would tweak this to be

```clojure
(defn- escape-script-tags [s]
  ...)
```

camsaul reviewed (approved):
Other than minor style suggestions this LGTM

👍

tlrobinson committed:
Merge branch 'master' of github.com:metabase/metabase into non-root-path-v2

tlrobinson committed:
Misc Clojure style

tlrobinson commented:
@metabase/core-developers Unless there are any objections I'm going to go ahead and merge this.

marcelmfs commented:
Is there any timeline for this feature to make to a release?

dncpax commented:
@marcelmfs I think it's included in 0.24.1, looking at the commits list it's there... on Commits on May 11, 2017

marcelmfs commented:
According to `https://github.com/metabase/metabase/compare/v0.24.1...master`, comparison, it's still on master. But made it's way to 0.25-rc already. So I might just wait for that, thanks!

dncpax commented:
Yep you're right I was reading it reversed...

wvengen commented:
Do I need to do anything to enable this? I've upgraded to 0.24.2, _Site URL_ shows the path, but I still see _"We're a little lost"_ on first access, and navigation bar doesn't show the path when clicking links.

marcelmfs commented:
I've solved this by using metabase-head docker container image.

jdelamar commented:
Hello. Does this feature work for you guys? I am still having some issues, but I may be missing something along the lines.

using `metabase/metabase-head` (or `metabase/metabase:v0.25.1`)

Here is my NGINX config:

```
upstream ctxbob_com_metabase {
        zone ctxbob_com_metabase.backend 64k;
	server 1.2.3.4:3000;
}
server {
        listen 80;
        server_name test.bob.com;
	location /metabase/ {
	    proxy_pass http://ctxbob_com_metabase;
	}
}
```

Using tcpdump, I can confirm that the metabase server receives the right URL as first message. On a brand new instance, tcp capture shows that the very first request is `http://test.bob.com/metabase/

However, metabase doesn't seem to register the prefix in the URI, and we see this in the logs
```
druid_metabase.1.2jftp4xvih54@debian-leader    | 08-03 14:52:54 INFO metabase.core :: Metabase Initialization COMPLETE
druid_metabase.1.2jftp4xvih54@debian-leader    | 08-03 14:53:56 INFO metabase.middleware :: Setting Metabase site URL to druid.aodtest.com

```
The end result is that every subsequent requests are sent without the `/metabase/` prefix
I can't really get to the admin page to fix the site URL. For instance, chrome sends this:
```
https://test.bob.com/app/dist/styles.bundle.css?393893ab6b24fff4707a542403400034
```
instead of
```
https://test.bob.com/metabase/app/dist/styles.bundle.css?393893ab6b24fff4707a542403400034
```

Is there anyway to either set the siteUrl via configuration or environment (I deploy in docker), or otherwise have the context root figured out on startup? Or I am fishing in the desert right now?

marcelmfs commented:
@jdelamar I think you need to use a regex and capture the URI to forward to the proxy.

```
location ^~ /metabase(/.*) {
   proxy_pass http://upstream$1;
}
```

Should do the trick.

jdelamar commented:
@marcelmfs, let me try this, but my TCP capture in my metabase container (i.e. after nginx) is confirming that my URL is proxied properly. Wireshark shows that the first request is a 
```
GET http://test.bob.com/metabase/
```
But that is the only request that makes it through. All the remaining requests from the webapp are sent without the prefix.

That matches what Chrome dev tools is showing. The first request is:
```
GET http://test.bob.com/metabase/
```
And the following are
```
https://test.bob.com/app/dist/styles.bundle.css?393893ab6b24fff4707a542403400034
```

This obviously gets canned at NGINX time, since nothing in the location is allowing be to trigger the upstream.

lfcipriani commented:
I deployed Metabase under /metabase path and I'm capturing on my load balancer logs a request that is being from Metabase frontend that are not inside the path:

`nginx.access.log: 10.0.1.237 - - [16/Aug/2017:12:24:53 +0000] "GET /app/img/dashboard_illustration.png HTTP/1.1" 404 18 "https://myhost/metabase/dashboards" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.12; rv:54.0) Gecko/20100101 Firefox/54.0"`

The install is working fine, only this request to `GET /app/img/dashboard_illustration.png` is leaking out of the path.

carneirotg commented:
Hello guys, I'm using Netflix OSS (Zuul Proxy and Eureka Service Discovery) and I can't get this to work. 

I'm also using mesos/marathon to scale my application. When I access it directly using the host:port that marathon gives me it works well:
Ex.:
`10.22.5.52:31252`

But when I register my app in Eureka and try to access it with Zuul, I can't pass through the login page. It keeps returning 'Unauthenticated', even after I enter with my credentials. Anyone has any idea?

Thank you!!

Digma commented:
@carneirotg Have you tried erasing your browser's cache after using the new url? Worked for me to remove the `Unauthenticated` issue

camlani commented:
@jdelamar Were you ever able to get this to work (or anyone else on that matter) ? I am still facing the same issue as it looks like you are. I am on the current version of Metabase 0.28.1 and have not been able to get the assets loaded properly with nginx.

Thanks for your help in advance! (I can access metabase still using its public facing ip and port just trying to use the below settings) Thanks!

NGINX Settings:

```
        listen 80 default_server;
        listen [::]:80 default_server;

       server_name sampledomain.com;

        location /metabase {
                proxy_pass http://localhost:3000;
        }
```
Metabase Admin Panel Site URL
http://sampledomain.com
or
http://sampledomain.com/metabase/

Navigate in Browser:
http://sampledomain.com/metabase/

Current issue :
```
GET http://sampledomain.com/app/dist/vendor.bundle.js?81977a256aef6526ffe7 net::ERR_ABORTED
(index):44 
GET http://sampledomain.com/app/dist/styles.bundle.js?81977a256aef6526ffe7 net::ERR_ABORTED
(index):44 
GET http://sampledomain.com/app/dist/app-main.bundle.js?81977a256aef6526ffe7 net::ERR_ABORTED
```

jdelamar commented:
No. I had to play games with NGINX to make metabase happy. Might I was
about to try again this week

Le dim. 11 févr. 2018 6:14 PM, Cameron <notifications@github.com> a écrit :

> @jdelamar <https://github.com/jdelamar> Were you ever able to get this to
> work (or anyone else on that matter) ? I am still facing the same issue as
> it looks like you are. I am on the current version of Metabase 0.28.1 and
> have not been able to get the assets loaded properly with nginx.
>
> Thanks for your help in advance! (I can access metabase still using its
> public facing ip and port just trying to use the below settings) Thanks!
>
> NGINX Settings:
>
>         listen 80 default_server;
>         listen [::]:80 default_server;
>
>        server_name sampledomain.com;
>
>         location /metabase {
>                 proxy_pass http://localhost:3000;
>         }
>
> Metabase Admin Panel Site URL
> http://sampledomain.com
> or
> http://sampledomain.com/metabase/
>
> Navigate in Browser:
> http://sampledomain.com/metabase/
>
> Current issue :
>
> GET http://sampledomain.com/app/dist/vendor.bundle.js?81977a256aef6526ffe7 net::ERR_ABORTED
> (index):44
> GET http://sampledomain.com/app/dist/styles.bundle.js?81977a256aef6526ffe7 net::ERR_ABORTED
> (index):44
> GET http://sampledomain.com/app/dist/app-main.bundle.js?81977a256aef6526ffe7 net::ERR_ABORTED
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/metabase/metabase/pull/4740#issuecomment-364799240>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AVDvfB0-SHoy0qjmQFv6foek3B7Vegu-ks5tT3RHgaJpZM4M4iq4>
> .
>


-- 
*Joel De Lamarre*

*Director R&D, *Network Performance Analytic
jdelamarre@accedian.com | +1.613.462.5399

-- 


Avis de confidentialité

Les informations contenues dans le présent message et dans toute pièce qui 
lui est jointe sont confidentielles et peuvent être protégées par le secret 
professionnel. Ces informations sont à l’usage exclusif de son ou de ses 
destinataires. Si vous recevez ce message par erreur, veuillez s’il vous 
plait communiquer immédiatement avec l’expéditeur et en détruire tout 
exemplaire. De plus, il vous est strictement interdit de le divulguer, de 
le distribuer ou de le reproduire sans l’autorisation de l’expéditeur. 
Merci.

Confidentiality notice

This e-mail message and any attachment hereto contain confidential 
information which may be privileged and which is intended for the exclusive 
use of its addressee(s). If you receive this message in error, please 
inform sender immediately and destroy any copy thereof. Furthermore, any 
disclosure, distribution or copying of this message and/or any attachment 
hereto without the consent of the sender is strictly prohibited. Thank you.

dncpax commented:
Hi there. Just a quick question: did you update site url? I was getting these links also wrong until I changed this setting.

camlani commented:
Hey @dncpax I have updated the site url on the admin settings page to both variants below with no luck, is there something else you did? I want to say that I am running the jar version of metabase, not the docker image, perhaps that is something as well. 
```
http://sampledomain.com
or
http://sampledomain.com/metabase/
```

jamsheedkv commented:
Hi @camlani, Did you get any solution for the above issue ?

mausch commented:
I'm also having issues getting this to work. I got as close as getting everything but the download links working (mentioned before [here](https://github.com/metabase/metabase/issues/2090#issuecomment-294107250)). 
To make things worse I'm running this on Kubernetes, meaning additional indirection everywhere. 
Maybe it could use a little documentation, if anyone could get it to work and could spare some time? 
For example I've found that I had to restart the app every time I changed the "Site URL" config, otherwise it would ignore the change.

Siddarth434 commented:
I am facing the same issue, not able make my downloads work. Using 0.30.2 version JAR. Please give some solution. Downloads are failing because it is not redirecting to the site URL configured in Nginx but my reports work like gem.

Siddarth434 commented:
> @jdelamar I think you need to use a regex and capture the URI to forward to the proxy.
> 
> ```
> location ^~ /metabase(/.*) {
>    proxy_pass http://upstream$1;
> }
> ```
> Should do the trick.

I tried and this is not working

Siddarth434 commented:
> @jdelamar Were you ever able to get this to work (or anyone else on that matter) ? I am still facing the same issue as it looks like you are. I am on the current version of Metabase 0.28.1 and have not been able to get the assets loaded properly with nginx.
> 
> Thanks for your help in advance! (I can access metabase still using its public facing ip and port just trying to use the below settings) Thanks!
> 
> NGINX Settings:
> 
> ```
>         listen 80 default_server;
>         listen [::]:80 default_server;
> 
>        server_name sampledomain.com;
> 
>         location /metabase {
>                 proxy_pass http://localhost:3000;
>         }
> ```
> Metabase Admin Panel Site URL
> http://sampledomain.com
> or
> http://sampledomain.com/metabase/
> 
> Navigate in Browser:
> http://sampledomain.com/metabase/
> 
> Current issue :
> 
> ```
> GET http://sampledomain.com/app/dist/vendor.bundle.js?81977a256aef6526ffe7 net::ERR_ABORTED
> (index):44 
> GET http://sampledomain.com/app/dist/styles.bundle.js?81977a256aef6526ffe7 net::ERR_ABORTED
> (index):44 
> GET http://sampledomain.com/app/dist/app-main.bundle.js?81977a256aef6526ffe7 net::ERR_ABORTED
> ```

Hi

Please try with the below nginx configuration

        listen 80 default_server;
        listen [::]:80 default_server;

       server_name sampledomain.com;

        location /metabase/ {
                proxy_pass http://localhost:3000/;
        }

jdelamar commented:
Hello, for reference, it has been working for us with the following:

```
server{

.... 
   upstream ctxdatahub2.npav.accedian.net__metabase_ {
      zone ctxdatahub2.npav.accedian.net__metabase__backend 64k;
      server 10.0.0.101:3000;

   }

  location /metabase/ {

             auth_request /auth;
             error_page 401 = @error401;


          proxy_pass http://ctxdatahub2.npav.accedian.net__metabase_/;
           proxy_redirect http://datahub2.npav.accedian.net https://datahub2.npav.accedian.net/metabase/;
        }
```
You can then access it using `https://bobserver.com/metabase`

Siddarth434 commented:
> Hello, for reference, it has been working for us with the following:
> 
> ```
> server{
> 
> .... 
>    upstream ctxdatahub2.npav.accedian.net__metabase_ {
>       zone ctxdatahub2.npav.accedian.net__metabase__backend 64k;
>       server 10.0.0.101:3000;
> 
>    }
> 
>   location /metabase/ {
> 
>              auth_request /auth;
>              error_page 401 = @error401;
> 
> 
>           proxy_pass http://ctxdatahub2.npav.accedian.net__metabase_/;
>            proxy_redirect http://datahub2.npav.accedian.net https://datahub2.npav.accedian.net/metabase/;
>         }
> ```
> You can then access it using `https://bobserver.com/metabase`

Hi @jdelamar ,

Are you able to download CSV from your domain? I am still not able to do so.

tlrobinson commented:
@Siddarth434 It looks like downloads are working in the query builder (`/question`) but not public questions (`/metabase/public/question`). Is that correct?

Siddarth434 commented:
Yes. It works when I am working on metabase environment. Fails when the report is embedded in my application.

tlrobinson commented:
@Siddarth434 PR to fix this here #8629 Please try it if possible.

Siddarth434 commented:
Thank you for fixing it. I will try this. Will this be fix be released with metabase next version Jar?
